Fix Fireworks AI rate limit exception mapping - detect "rate limit" text in error messages

colesmcintosh opened the pull request:
## Title

Fix Fireworks AI rate limit exception mapping - detect "rate limit" text in error messages

## Relevant issues

Fixes the issue where Fireworks AI 429 rate limit errors were being incorrectly translated to 400 BadRequestError status codes instead of preserving the rate limit semantics.

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🐛 Bug Fix

## Changes

### Problem
Fireworks AI was returning rate limit errors with the message "rate limit exceeded, please try again later" but these were being mapped to `BadRequestError` (400 status) instead of `RateLimitError` (429 status). This happened because:

1. Fireworks AI returns `{"error":{"type":"invalid_request_error","message":"rate limit exceeded, please try again later"}}` 
2. The existing rate limit detection only checked for the literal string "429" in the error message
3. Since the error contained "invalid_request_error", it was caught by the BadRequestError condition before reaching proper rate limit detection

### Solution
**Enhanced rate limit detection** in `litellm/litellm_core_utils/exception_mapping_utils.py`:
- Modified the rate limit check from `if "429" in error_str:` to `if "429" in error_str or "rate limit" in error_str.lower():`
- This now detects rate limits both by status code AND by text content
- Preserves existing behavior while fixing the Fireworks AI case

### Testing
**Added comprehensive test coverage** in `tests/local_testing/test_exceptions.py`:
- `test_fireworks_ai_rate_limit_text_detection()` - Tests the exact real-world scenario with "rate limit exceeded" message
- Verifies that text-based rate limit detection works correctly
- Confirms that the fix doesn't break existing functionality

### Impact
- Fireworks AI rate limit errors now correctly map to `RateLimitError` with proper 429 semantics
- All OpenAI-compatible providers benefit from improved rate limit detection
- Backward compatibility maintained - existing "429" detection still works
- No breaking changes to existing functionality

### Test Results
<img width="680" alt="Screenshot 2025-06-05 at 3 47 19 PM" src="https://github.com/user-attachments/assets/f00eb0d3-2cba-4d5a-afa3-c3f41f0ef27d" />

This fix ensures that users will now receive proper `RateLimitError` exceptions when Fireworks AI rate limits are hit, allowing for appropriate retry logic and error handling.

colesmcintosh committed:
Enhance exception mapping for Fireworks AI: add better handling for 429 status codes and text-based rate limit detection. Update tests to verify correct mapping to RateLimitError for both 429 and related error messages.

ishaan-jaff commented on litellm/litellm_core_utils/exception_mapping_utils.py:
```diff
@@ -274,7 +274,7 @@ def exception_type(  # type: ignore  # noqa: PLR0915
                     )
 
-                if "429" in error_str:
+                if "429" in error_str or "rate limit" in error_str.lower():
```
working on improving code quality in this file. Can we have a class called ExceptionCheckers and then add a static method that checks is_error_str_rate_limit()

ishaan-jaff replied:
also can you confirm with a manual test to fireworks that we end up mapping it correctly ?

ishaan-jaff reviewed (changes requested):
mostly lgtm, 1 change

colesmcintosh committed:
Add ExceptionCheckers class for improved error string detection

Introduce the ExceptionCheckers class to encapsulate methods for checking error conditions in exception strings, specifically for identifying rate limit errors. Update the Fireworks AI exception mapping tests to cover various scenarios, including standard 429 errors and text-based detection, ensuring accurate mapping to RateLimitError. Enhance test coverage for both positive and negative cases of rate limit detection.

colesmcintosh commented on tests/local_testing/test_exceptions.py:
```diff
@@ -819,70 +819,96 @@ def test_exception_mapping(provider):
+        "Rate limit exceeded, please try again later", 
+        "RATE LIMIT ERROR",
+        "Error 429: rate limit",
+        '{"error":{"type":"invalid_request_error","message":"rate limit exceeded, please try again later"}}',
```
@ishaan-jaff Here is the manual check on the exact error from Fireworks

ishaan-jaff reviewed (approved):
LGTM, thanks !

colesmcintosh commented:
@ishaan-jaff do you think it would be worth it to update this to do regex matching for messages like "too many requests" as well?

krrishdholakia commented on tests/local_testing/test_exceptions.py:
```diff
@@ -819,6 +819,98 @@ def test_exception_mapping(provider):
     pass
 
 
+def test_fireworks_ai_exception_mapping():
```
in here - https://github.com/BerriAI/litellm/tree/main/tests/test_litellm/litellm_core_utils
